frontends/ast/simplify.cc

```diff
@@ -2867,12 +2865,76 @@ bool AstNode::simplify(bool const_fold, int stage, int width_hint, bool sign_hin
 		{
 			// big case block
 
+			int stride = 1;
+
+			int case_width_hint;
+			bool case_sign_hint;
+			shift_expr->detectSignWidth(case_width_hint, case_sign_hint);
+
+			// Extract (index)*(width) from non_opt_range pattern ((@selfsz@((index)*(width)))+(0)).
+			AstNode *lsb_expr =
+				shift_expr->type == AST_ADD && shift_expr->children[0]->type == AST_SELFSZ &&
+				shift_expr->children[1]->type == AST_CONSTANT && shift_expr->children[1]->integer == 0 ?
+				shift_expr->children[0]->children[0] :
+				shift_expr;
+
+			// Optimization: Extract stride from indexing of two-dimensional packed arrays and
+			// variable slices on the form dst[i*stride +: width] = src.
+			if (lsb_expr->type == AST_MUL &&
+			    (lsb_expr->children[0]->type == AST_CONSTANT ||
+			     lsb_expr->children[1]->type == AST_CONSTANT))
+			{
+				int const_i = lsb_expr->children[1]->type == AST_CONSTANT;
+				stride = (int)lsb_expr->children[const_i]->integer;
+			}
+			else if (member_node)  // Member in packed struct/union
+			{
+				// Clamp chunk to range of member within struct/union.
+				log_assert(!wire_offset && !children[0]->id2ast->range_swapped);
+
+				// When the (* nowrshmsk *) attribute is set, a CASE block is generated below
+				// to select the indexed bit slice. When a multirange array is indexed, the
+				// start of each possible slice is separated by the bit stride of the last
+				// index dimension, and we can optimize the CASE block accordingly.
+				// The dimension of the original array expression is saved in the 'integer' field.
+				int dims = children[0]->integer;
+				stride = wire_width;
+				for (int dim = 0; dim < dims; dim++) {
+					stride /= get_struct_range_width(member_node, dim);
+				}
+			}
+
+			// For (truncated) i*stride to be within the range of dst, the following must hold:
+			//   i*stride ≡ bitno (mod shift_mod), i.e.
+			//   i*stride = k*shift_mod + bitno
+			//
+			// The Diophantine equation on the form ax + by = c:
+			//   stride*i - shift_mod*k = bitno
+			// has solutions iff c is a multiple of d = gcd(a, b), i.e.
+			//   bitno mod gcd(stride, shift_mod) = 0
+			//
+			// long long is at least 64 bits in C++11
+			int max_bits = case_width_hint;
+			long long max_offset = (1ll << (max_bits - case_sign_hint)) - 1;
+			long long min_offset = case_sign_hint ? -(1ll << (max_bits - 1)) : 0;
+			long long shift_mod = 1ll << (case_width_hint - case_sign_hint);
+			// std::gcd requires C++17
+			// long long bitno_div = std::gcd(stride, shift_mod);
+			long long bitno_div = gcd((long long)stride, shift_mod);
```

Comment:
I feel like this undoes much of the optimisation when `bitno_div` can only be power-of-2, not that I don't see why it's done to address the overflown cases.

Reply:
Well, it's not ideal, although power-of-2 will perfectly optimize common (most?) cases (e.g. `gcd(8, 256) = 8`), and partially optimize others (e.g. `gcd(12, 256) = 4`).

Since we know the value of `stride`, it might be possible to identify cases where the index expression cannot possibly overflow, and fully optimize those cases. I'll look into it.

Reply:
How about now? :sweat_smile: 

All tests still pass with a forced `use_case_method = true`, and the results from running the tests in `shift_issue3.zip` above are:
```
   Chip area for module '\reg_demux_nowrshmsk_mdim_arr': 19213.891200
   Chip area for module '\reg_demux_nowrshmsk_typedef': 19115.271000
   Chip area for module '\reg_demux_pow2': 5513.961600
   Chip area for module '\reg_demux': 21857.661000
```

As far as I can tell, the PR now only brings improvements, without any regressions.